src/lib/io/sol-ipm.c

```diff
@@ -0,0 +1,370 @@
+    const void *data;
+    struct sol_blob *last_remote_blob; /* Hackish... I store last blob received here, to avoid having yet another list to store them */
+};
+
+struct sol_ipm_consumed_handler {
+    uint32_t id;
+    void (*message_consumed_cb)(void *data, uint32_t id, struct sol_blob *message);
+    const void *data;
+};
+
+static struct sol_vector ipm_receivers = SOL_VECTOR_INIT(struct sol_ipm_receiver_entry);
+static struct sol_vector ipm_consumed_handlers = SOL_VECTOR_INIT(struct sol_ipm_consumed_handler);
+static struct device *ipm_sender, *ipm_receiver;
+static uint32_t ipm_max_id;
+static bool initialised;
+
+static void
+free_remote_blob(struct sol_blob *blob)
+{
+    int r;
+
+    r = ipm_send(ipm_sender, 1, SOL_IPM_ID_FREE_REMOTE, &blob,
+        sizeof(uintptr_t));
+    if (r != 0) {
+        /* Should I care if return was EBUSY?*/
+        SOL_WRN("Could not send consumed message to remote core: %d", r);
+    }
+}
+
+static void
+free_ipm_blob(struct sol_blob *blob)
+{
+    struct sol_blob_ipm_received *ipm_blob = (struct sol_blob_ipm_received *)blob;
+
+    free_remote_blob(ipm_blob->remote);
+
+    /* Free local shadow */
+    free(blob);
+}
+
+extern unsigned char *heap_base;
+extern unsigned int heap_sz, heap_max;
+
+static void
+ipm_receiver_process(void *data)
+{
+    int idx;
+    struct sol_blob_ipm_sent *message = data;
+    struct sol_ipm_receiver_entry *receiver;
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_receivers, receiver, idx) {
+        if (receiver->id == message->id) {
+            struct sol_blob_ipm_received *remote_blob;
+
+            /* Setup local shadow blob whose `mem` points to the same `mem` as remote one*/
+            remote_blob = calloc(1, sizeof(struct sol_blob_ipm_received));
+            if (!remote_blob) {
+                SOL_WRN("Could not create blob to deliver IPM message");
+                goto fail;
+            }
+
+            sol_blob_setup(&remote_blob->base, &SOL_BLOB_TYPE_IPM_RECEIVED,
+                message->base.mem,
+                message->base.size);
+
+            remote_blob->remote = &message->base;
+            receiver->receive_cb((void *)receiver->data, message->id,
+                &remote_blob->base);
+            return;
+        }
+    }
+
+    SOL_INF("Processed IPM id %" PRIu32 " but no receiver was found!",
+        message->id);
+
+fail:
+    /* Let's unref this blob on remote */
+    free_remote_blob(&message->base);
+}
+
+static void
+ipm_reaper_process(void *data)
+{
+    int idx;
+    struct sol_blob_ipm_sent *message = data;
+    struct sol_ipm_consumed_handler *handler;
+
+    SOL_VECTOR_FOREACH_IDX (&ipm_consumed_handlers, handler, idx) {
+        if (handler->id == message->id) {
+            handler->message_consumed_cb((void *)handler->data, message->id,
+                message->base.parent);
+            sol_blob_unref(&message->base);
+            free(message); /* This blob is SOL_BLOB_TYPE_NO_FREE, so we need to
```

Comment:
the `NO_FREE` doesn't free the pointed memory, but it does free the blob structure. Actually this is something I've asked @cabelitos to review, but AFAIU it was not done yet. See https://github.com/solettaproject/soletta/pull/1918#discussion_r60498120